### PR TITLE
[RepositoryManager] Move towards closure based async APIs

### PR DESCRIPTION
**Pull request by aciidb0mb3r**

### CODE REVIEW

File: Sources/SourceControl/RepositoryManager.swift [hunk 1]
```diff
@@ -165,54 +121,74 @@ public class RepositoryManager {
...
-    /// necessary, and immediately return. The client can add observers to the
-    /// result in order to know when the repository is available.
-    public func lookup(repository: RepositorySpecifier) -> RepositoryHandle {
-        // Check to see if the repository has been provided.
-        if let handle = repositories[repository.url] {
-            // Fetch and update the repository when it is being looked up.
-            // We need some mechanism to maybe not do this for pinned repos.
+    /// necessary.
+    public func lookup(repository: RepositorySpecifier, completion: @escaping (Result<RepositoryHandle, AnyError>) -> Void) {
+        DispatchQueue.global().async {
```
Comment (aciidb0mb3r) [inline Sources/SourceControl/RepositoryManager.swift]: This area is WIP

File: Sources/SourceControl/RepositoryManager.swift [hunk 2]
```diff
@@ -140,6 +93,9 @@ public class RepositoryManager {
     // specifier but then is capable of efficiently determining if two
     // repositories map to the same location.
     private var repositories: [String: RepositoryHandle] = [:]
+
+    /// Lock to protect concurrent reads and mutations to repositories registory.
+    private var repostoriesLock = Lock()
```
Comment (ddunbar) [inline Sources/SourceControl/RepositoryManager.swift]: Sic (repostories)

File: Sources/SourceControl/RepositoryManager.swift [hunk 3]
```diff
@@ -165,54 +121,74 @@ public class RepositoryManager {
...
+    /// necessary.
+    public func lookup(repository: RepositorySpecifier, completion: @escaping (Result<RepositoryHandle, AnyError>) -> Void) {
+        DispatchQueue.global().async {
+            repostoriesLock.whileLocked {
+                // Check to see if the repository has been provided.
+                if let handle = self.repositories[repository.url] {
+                    // Fetch and update the repository when it is being looked up.
+                    // We need some mechanism to maybe not do this for pinned repos.
+                    do {
+                        let repo = try handle.open()
```
Comment (ddunbar) [inline Sources/SourceControl/RepositoryManager.swift]: This is holding the `repositoriesLock` while doing the expensive operation, which would prevent it from every being particularly concurrent.

This is one reason we need something like the handle to be able to be placed into the repositories mapping _while_ it is being fetched (since if someone else fetches it they need to just chain onto that same fetch operation). We might be able to foist this onto the client and say "don't clone the same URL multiple times" with this API, but someone has to handle managing this, and it seems ideal to keep it contained.

File: Sources/SourceControl/RepositoryManager.swift [hunk 4]
```diff
@@ -165,54 +121,74 @@ public class RepositoryManager {
...
+                        try repo.fetch()
+                    } catch {
+                        // FIXME: Better error handling.
+                        // Maybe just print a warning that we couldn't fetch it
+                        // since we already have the repo.
+                        print("Error while fetching the repo: \(handle)")
```
Comment (ddunbar) [inline Sources/SourceControl/RepositoryManager.swift]: Why wouldn't the result in this case be the error?

File: Sources/SourceControl/RepositoryManager.swift [hunk 5]
```diff
@@ -165,54 +121,74 @@ public class RepositoryManager {
...
+                }
+            }
+
+            // Ensure nothing else exists at the subpath.
+            let repositoryPath = self.path.appending(subpath)
+            if localFileSystem.exists(repositoryPath) {
+                _ = try? removeFileTree(repositoryPath)
+            }
+
+            let result: Result<RepositoryHandle, AnyError>
```
Comment (ddunbar) [inline Sources/SourceControl/RepositoryManager.swift]: This could use `Result.init(_ body: () throws -> Value)` to be simpler, I think.

File: Sources/SourceControl/RepositoryManager.swift (hunk 4 above)
Comment (aciidb0mb3r) [inline Sources/SourceControl/RepositoryManager.swift]: We still have the repo, we just werent able to update it, that should just be a warning right?

Comment (ddunbar) [inline Sources/SourceControl/RepositoryManager.swift]: I'm not sure, it depends on the client... as phrased the API is "give me the up-to-date repo", and we failed at that. I think we could throw an error that communicates this specific information, and clients could downgrade from an error if they want to.

**Comment by aciidb0mb3r** [comment]
@ddunbar I need to add some edge cases tests but this PR can be reviewed otherwise

File: Sources/SourceControl/RepositoryManager.swift [hunk 6]
```diff
@@ -140,6 +122,9 @@ public class RepositoryManager {
     // specifier but then is capable of efficiently determining if two
     // repositories map to the same location.
     private var repositories: [String: RepositoryHandle] = [:]
+
+    /// Lock to protect concurrent reads and mutations to repositories registory.
+    private var repositoriesLock = Lock()
```
Comment (ddunbar) [inline Sources/SourceControl/RepositoryManager.swift]: Should we prefer a queue versus a lock? I like the queue pattern for serialized access to state, even though they are basically the same.

File: Sources/SourceControl/RepositoryManager.swift [hunk 7]
```diff
@@ -165,54 +150,127 @@ public class RepositoryManager {
...
-            // We need some mechanism to maybe not do this for pinned repos.
-            do {
-                let repo = try handle.open()
-                try repo.fetch()
-            } catch {
-                // FIXME: Better error handling.
-                print("Error while fetching the repo: \(handle)")
+    /// necessary.
+    public func lookup(repository: RepositorySpecifier, completion: @escaping (LookupResult) -> Void) {
+        DispatchQueue.global().async {
```
Comment (ddunbar) [inline Sources/SourceControl/RepositoryManager.swift]: IMHO we might as well break out our own parallel queue to use for this, vs the global queue.

File: Sources/SourceControl/RepositoryManager.swift [hunk 8]
```diff
@@ -165,54 +150,127 @@ public class RepositoryManager {
...
+            // Get the handle for this repository.
+            let handle = self.getHandle(repository: repository)
+            switch handle.status {
```
Comment (ddunbar) [inline Sources/SourceControl/RepositoryManager.swift]: This doesn't seem safe, the concurrent access to `handle.state` isn't protected by anything.

As a general rule, if you have some variable protected by a lock, it must *always* be accessed via that protection (google for why double checked locking isn't safe if you are curious). We need to find a pattern where the state is more protected.

File: Sources/SourceControl/RepositoryManager.swift [hunk 9]
```diff
@@ -165,54 +150,127 @@ public class RepositoryManager {
...
+
+            case .available:
+                let updatedResult = try! LookupResult {
```
Comment (ddunbar) [inline Sources/SourceControl/RepositoryManager.swift]: why `try!`?

File: Sources/SourceControl/RepositoryManager.swift [hunk 10]
```diff
@@ -165,54 +150,127 @@ public class RepositoryManager {
...
+                    return handle
+                }
+                completion(updatedResult)
+
+            case .pending:
+                self.repositoriesLock.withLock {
+                    // Make sure we're still in pending status after aquiring this lock.
+                    guard handle.status == .pending else {
+                        // FIXME: This is wrong if handle is in error state.
+                        completion(Result(handle))
```
Comment (ddunbar) [inline Sources/SourceControl/RepositoryManager.swift]: Calling the completion handler with the lock held is very dangerous.

File: Sources/SourceControl/RepositoryManager.swift [hunk 11]
```diff
@@ -165,54 +150,127 @@ public class RepositoryManager {
...
+
+                // Call the completion and any previous completion handlers we queued.
+                completion(result)
+                handle.completionHandler?(result)
+                // Release the handlers.
+                self.repositoriesLock.withLock {
+                    handle.completionHandler = nil
+                }
+            }
+        }
```
Comment (ddunbar) [inline Sources/SourceControl/RepositoryManager.swift]: I suspect this code can be simplified if you first always get the handle, and chain the completion handler, and then act on the state of the handle. That should eliminate the duplicate places the completion handler is called.

File: Sources/SourceControl/RepositoryManager.swift [hunk 12]
```diff
@@ -165,54 +150,127 @@ public class RepositoryManager {
...
+                    fatalError("unable to save manager state \(error)")
+                }
+
+                // Call the completion and any previous completion handlers we queued.
+                completion(result)
+                handle.completionHandler?(result)
+                // Release the handlers.
+                self.repositoriesLock.withLock {
+                    handle.completionHandler = nil
+                }
```
Comment (ddunbar) [inline Sources/SourceControl/RepositoryManager.swift]: This isn't safe I don't think. What needs to happen is to take the handler out and nil the handles entry with the lock held, then call it. If a client sends a request while the handler is running above, it will be lost in this current code.

File: Sources/SourceControl/RepositoryManager.swift (hunk 9 above)
Comment (aciidb0mb3r) [inline Sources/SourceControl/RepositoryManager.swift]: This is Result's closure init which can throw but since the error here is `AnyError`, it will never fail

Comment (ddunbar) [inline Sources/SourceControl/RepositoryManager.swift]: Hrm, that is annoyingly subtle. I wonder if there is some way to have a specific init for the AnyError case which downgrades it from throwing?

**Comment by aciidb0mb3r** [comment]
@ddunbar ready for rereview

File: Sources/SourceControl/RepositoryManager.swift [hunk 13]
```diff
@@ -164,55 +159,120 @@ public class RepositoryManager {
...
-                try repo.fetch()
-            } catch {
-                // FIXME: Better error handling.
-                print("Error while fetching the repo: \(handle)")
+    /// This will initiate a clone of the repository automatically, if necessary.
+    public func lookup(repository: RepositorySpecifier, completion: @escaping LookupCompletion) {
+        concurrentQueue.async {
+            // Get the handle and status. The completion will be stored or chained inside handle.
+            let (handle, status) = self.getHandle(repository: repository, lookupCompletion: completion)
+            let result: LookupResult
```
Comment (ddunbar) [inline Sources/SourceControl/RepositoryManager.swift]: This code is still unprotected, and can race with other clients (e.g., two clients could trigger a fetch simultaneously.

File: Sources/PackageGraph/RepositoryPackageContainerProvider.swift [hunk 14]
```diff
@@ -39,24 +39,7 @@ public class RepositoryPackageContainerProvider: PackageContainerProvider {
...
-            }
-        }
-        while wasAvailableCondition.whileLocked({ wasAvailableOpt == nil}) {
-            wasAvailableCondition.wait()
-        }
-        let wasAvailable = wasAvailableOpt!
-        if !wasAvailable {
-            throw RepositoryPackageResolutionError.unavailableRepository
-        }
+        let handle = try repositoryManager.lookupSynchronously(repository: identifier)
```
Comment (ddunbar) [inline Sources/PackageGraph/RepositoryPackageContainerProvider.swift]: BTW, I would land this change independently of other parts.

File: Sources/SourceControl/RepositoryManager.swift [hunk 15]
```diff
@@ -145,53 +184,93 @@ public class RepositoryManager {
...
+                let newHandle = RepositoryHandle(manager: self, repository: repository, subpath: subpath)
+                self.repositories[repository.url] = newHandle
+                handle = newHandle
+            }
+            return handle
+        }
+    }
+
+    /// Fetches the repository of a handle in pending status and returns the result.
+    private func fetchRepository(of handle: RepositoryHandle) -> LookupResult {
```
Comment (ddunbar) [inline Sources/SourceControl/RepositoryManager.swift]: This method should document it is expected to be called protected by the internal queue.

File: Sources/SourceControl/RepositoryManager.swift [hunk 16]
```diff
@@ -37,6 +39,35 @@ public class RepositoryManager {
...
+                case .available:
+                    break
+                case .error:
+                    if status == .pending {
+                        self = status
+                        return true
+                    }
+                }
+                return false
+            }
```
Comment (ddunbar) [inline Sources/SourceControl/RepositoryManager.swift]: I think this method can be inlined now and it will make PR more clear.

File: Sources/SourceControl/RepositoryManager.swift [hunk 17]
```diff
@@ -54,6 +85,8 @@ public class RepositoryManager {
         /// The status of the repository.
         fileprivate var status: Status = .uninitialized
 
+        fileprivate let serialQueue = DispatchQueue(label: "org.swift.swiftpm.repohandle-serial")
```
Comment (ddunbar) [inline Sources/SourceControl/RepositoryManager.swift]: Comment plz!!! This is critical stuff, this is a good place to document what the strategy is, what it is protecting, what to watch out for, etc.

File: Sources/SourceControl/RepositoryManager.swift [hunk 18]
```diff
@@ -145,53 +155,94 @@ public class RepositoryManager {
...
+                self.repositories[repository.url] = newHandle
+                handle = newHandle
+            }
+            return handle
+        }
+    }
+
+    /// Fetches the repository of a handle in pending status and returns the result.
+    /// precondition: The handle must be in uninitialized or error state.
+    private func fetchRepository(of handle: RepositoryHandle) -> LookupResult {
```
Comment (ddunbar) [inline Sources/SourceControl/RepositoryManager.swift]: Given that this can only be used on the locked queue... I think it might be most obvious & safest to simply inline that there, and then move the function that executes on the handle queue out to its own place, and then document that it *must* be run on the handle's internal queue.

**Comment by aciidb0mb3r** [comment]
@ddunbar updated

File: Sources/SourceControl/RepositoryManager.swift [hunk 19]
```diff
@@ -145,53 +157,84 @@ public class RepositoryManager {
...
+            let handle = self.getHandle(repository: repository)
+            // Dispatch the action we want to take on the serial queue of the handle.
+            handle.serialQueue.async {
+                let result: LookupResult
+                switch handle.status {
+                case .available:
+                    // FIXME: Need to only do this for first lookup.
+                    result = try! LookupResult {
+                        // Fetch and update the repository when it is being looked up.
+                        // We need some mechanism to maybe not do this for pinned repos.
```
Comment (ddunbar) [inline Sources/SourceControl/RepositoryManager.swift]: FYI, this mechanism wouldn't be here: it would be on the resolver to not ask for pinned repos, I think.

**Comment by aciidb0mb3r** [comment]
@swift-ci please test
